(DOCSP-27564) Add documentation for sync migrations (#396)
mongodb/docs-app-services: source/sync/migrate-sync-modes.txt

.. _realm-sync-migrate-modes:

=============================================
Migrate Partition-Based Sync to Flexible Sync
=============================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

You can automatically migrate App Services Apps that use Partition-Based Sync
to Flexible Sync. We recommend doing this, as Partition-Based Sync is an older
mode.

Other than upgrading the SDK version, migrating sync modes doesn't require any 
changes to your client app code. Your App's backend will automatically translate
the client's Partition-Based Sync requests into Flexible Sync.

> .. QUESTION: is there a performance hit for the translation?

Reviewer: nope! there is a one-time cost (roughly equivalent to the regular cost of a client reset) when the PBS client first tries to connect, but once that handshake happens it will be a regular flexible sync client moving forward

Migrating results in Partition-Based Sync clients communicating with a
Flexible Sync App Services App. You cannot reverse this and have Flexible Sync
clients communicate with a Partition-Based Syn App Services App.

Requirements
------------

- Your App must currently use Partition-Based Sync
- Linked Atlas cluster must be running mongodb 5.0+ for Flexible Sync compatibility

> - Linked Atlas cluster must have time-based oplog setting enabled

Reviewer: this carries with it two other implicit requirements:
- linked cluster must have storage auto-scaling enabled
- linked cluster must be M10+ to access this setting

Author: Ooh, those are good details. I remember you mentioning auto-scaling, but I didn't know you need M10+ to access it.

I'll add these to make them clear to readers.

- realm-core vTBD+, which means these SDK versions: TBD

.. _realm-sync-migrate-what-to-know:

> What to Know Before migrating

Reviewer: another important callout here is that old clients will get a client reset, so they should make sure they have a client reset handler mode that makes sense for them (and maybe worth encouraging them to use client reset with recovery if they don't already)

	 Clients can receive updates to MongoDB data.

#. Initializing flexible sync
   
   Clients cannot connect at this stage. The time this takes is directly
   proporional to the amount of data in synced collections.

#. Migration completed

   **Manage sync migration** appears on Sync page if the migration hasn't been 
   committed or rolled back. This button lets admins commit or roll back manually.

Migrate Partition-Based Sync App to Flexible Sync
-------------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can roll it back 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

During migration, client apps may experience side effects, most of which are similar
to what happens if you disable and re-enable sync. These are some side effects

> of migration:

Author: Great point! I'll clarify that these are additional side effects.

- In the App Services UI, the Rules and Schemas pages will be unresponsive.
- Your App Services Appp will experience a period of read-only sync and then a 
  period of downtime.
- If Partition-Based Sync default client max offline time is unset, the new
  Flexible Sync config will set it to 30 days.
- Storage in your Atlas cluster used by sync is temporarily doubled. Partition-Based
  Sync and Flexible Sync must temporarily co-exist. Depending on the size of your
  synced data, this could affect billing.

.. QUESTION: If someone were to cancel a migration - or roll back a migration,
   they'd potentially be on the hook for costs associated with subsequent 

>    migration attempts, correct?

Reviewer: what costs are you referring to here?

Author: Ah, sorry. Potential costs related to the doubling sync metadata. It's doubled every time because they have to start over, right?

Reviewer: Ah I see. Canceling/rolling back a migration automatically clears the flexible sync metadata - so their storage should go back to what it was prior to starting the migration

To migrate from Partition-Based Sync to Flexible Sync:

#. In the App Services UI, go to your App and select the Sync page.
#. At the top of the Sync page, select :guilabel:`[button-text]`.

Your App will lose the migration in the following cases:

- Partition-Based Sync is re-enabled
- You roll back the migration

Migrate permissions
```````````````````

Your App's sync permissions will be automatically migrated if the Partition-Based

> Sync permissions have direct Flexible Sync equivalents.

Reviewer: Worth mentioning that in this case, it will override any app services rules that were previously defined

Some Partition-Based Sync permission strategies can't translate directly to Flexible 
Sync. You need to manually migrate permissions in these situations:

- Use of ``%function``
	- Can't be used in FBS document filters
	- Most commonly used with the ``%partition`` expansion, which doesn't have an FBS equivalent

- Use of a recursing operator (``%or``, ``%not``, ``%nor``, ``%and``)
	- Technically works, but could cause issues.
	- Users should test these and check if they need to make changes to the auto-migrated rules.
	- Testing new permissions won't work on existing app. Need a new app to test manually-migrated permissions rules.

Refer to the :ref:`Partiton-Based Sync <pbs-rule-migration>` reference for examples
of how to map these Partition-Based Sync permissions to Flexible Sync permissions.

Commit Migration
~~~~~~~~~~~~~~~~

> After your migration has completed, it is in a non-permanent state. You can manually

Reviewer: [nit] not a huge fan of "non-permanent" here since it may imply that we would roll it back without manual intervention

Author: Yeah, I don't think it's quite right. I've updated to `an evaluation state`. Do you think that conveys the purpose of the roll back window?

commit the migration to make it permanent or roll back the migration to return your
App to Partiton-Based Sync.

If you don't commit or roll back the migration, it will automatically commit when
the PBS translator falls off of the oplog.

> .. QUESTION: What will the oplog show, exactly, for PBS translator

Reviewer: This is directly proportional to the cluster's time-based oplog setting: https://www.mongodb.com/docs/atlas/cluster-additional-settings/#set-minimum-oplog-window

e.g. if they set the value to 24 hours, then they can assume that they have ~24 hours to roll back starting from when the migration _started_ (not ended!)

while the rollback window is open, they can continue to bump that value to expand their window at the cost of additional cluster storage

and you cannot roll back the migration.

.. TODO: Make this more specific

To manually commit a migration:

#. In the App Services UI, go to your App and select the Sync page.
#. At the top of the Sync page, select :guilabel:`[button-text]`.

Cancel Migration
~~~~~~~~~~~~~~~~

You can cancel a migration at any time while it is in progress. This will return
your App the Partition-Based Sync, with all settings as they were before the
migration started.

If you cancel a migration, you need to start over from the beginning with future
migration attempts.

.. TODO: Make this more specific

To cancel a migration:

#. In the App Services UI, go to your App and select the Sync page.
#. At the top of the Sync page, select :guilabel:`[button-text]`.

Roll Back Migration
~~~~~~~~~~~~~~~~~~~

You can roll back a migration after it has completed and before it has been committed.

Completed migrations are automatically committed when the PBS translator falls 
off of the oplog.
.. QUESTION: What will the oplog show, exactly, for PBS translator

You can increase or decrease the migration roll back window by adjusting
oplog retentention.

> .. TODO: Add link to oplog retention docs

Reviewer: here ya go - https://www.mongodb.com/docs/atlas/cluster-additional-settings/#set-minimum-oplog-window

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

Migrating an App that uses Partition-Based Sync to Flexible Sync is an automatic
process. Other than upgrading the SDK version, migrating sync modes doesn't require 
any changes to your client app code. Your App's backend will automatically translate
the client's Partition-Based Sync requests into Flexible Sync.

Migrating results in Partition-Based Sync clients communicating with a
Flexible Sync Atlas App Services App.

Requirements
------------

Your App Services App:

- Must currently use Partition-Based Sync

Linked Atlas Cluster:

- Must be running mongodb 5.0+ for Flexible Sync compatibility
- Must have storage auto-scaling enabled
- Must be M10+ to access this setting
- Must have time-based oplog setting enabled

Client Apps:

> - ``realm-core`` vTBD+, which means these SDK versions: TBD

Author: @kmorkos, do we know what version of realm-core and SDK versions users will need?

Reviewer: Not yet, tagging @michael-wb since he's currently working on the realm-core side

your App's backend. You should be aware of the side effects and plan ahead.

How long a migration takes is directly proportional to the amount of data that
needs to be migrated. The more data, the longer it will take to migrate.

Migrations are not permenant immediately after completion. They can be canceled
during the migration or rolled back within a certain timeframe after the migration
is finished.

Progress notifications will not work after migration because Flexible Sync does
not yet support them.

Client apps that use an SDK with with a realm-core version lower than TBD will
experience a client reset after migration. You should use a 
:ref:`client reset <handle-a-client-reset>` handler with recovery enabled.

.. _realm-sync-migrate-stages:

Migration Stages
~~~~~~~~~~~~~~~~

There are three stages to a Sync mode migration. You can see your migration's
stage in the App Services UI.

#. Disabling Partition-Based Sync
   
   At this stage, client apps can connect to your App's backend, but local writes 
   will not sync. These writes will be lost if you don't have a 
   :ref:`client reset handler <handle-a-client-reset>` to recover unsynced changes.
	 

>    Clients can also receive updates to MongoDB data, but they won't

Author: @kmorkos, "MongoDB data" in this instance refers to the linked cluster, correct?

Reviewer: correct

   be able to recieve upload acknowledgements.

#. Initializing Flexible Sync
   
   Clients cannot connect at this stage. The time this takes is directly
   proportional to the amount of data in synced collections.

#. Migration completed

   **Manage sync migration** appears on Sync page if the migration hasn't been 
   committed or rolled back. This button lets admins commit or roll back manually.

Migrate Partition-Based Sync App to Flexible Sync
-------------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can roll it back 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

During migration, client apps may experience side effects, most of which are similar
to what happens if you disable and re-enable sync. These are some additional
side effects of migration:

> - In the App Services UI, the Rules and Schemas pages will be unresponsive.

Author: @kmorkos, do you think "unresponsive" is a fair description of what will happen? I haven't seen that side effect.

Reviewer: It wouldn't be unresponsive, just read-only. So they can still see their schemas (I think there's a bug that might lead to a broken view of rules), but if they try to make any changes to their config it will fail

Author: Ah, that makes a lot more sense. Thank you!

   :guilabel:`Manage Sync Migration` appears on the Device Sync page if the 
   migration hasn't been committed or reverted.

Migrate Partition-Based Sync to Flexible Sync
---------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

During migration, client apps may experience side effects, most of which are similar
to what happens if you :ref:`terminate <terminating-realm-sync>` and re-enable
Device Sync. These are some additional side effects of migration:

- In the App Services UI, the Rules and Schema pages will be read-only.
- Your App will experience a period of read-only Sync and then a 
  period of downtime.
- If Partition-Based Sync :ref:`client max offline time <client-maximum-offline-time>`

>   is unset, the new Flexible Sync configuration will set it to 30 days.

Reviewer: UPDATE: we're removing the "client max offline time" setting for PBS apps, so we decided that for a migration, we will always set the flexible sync max offline time to 30 days regardless of their old PBS setting

   
   .. image:: /images/begin-migration.png
      :alt: Begin a migration
   
#. Read the information about migrating, then select 
   :guilabel:`Next: Migration Requirements`.
#. Ensure you meet the migration requirements, the select 
   :guilabel:`Start Migration`.

After a migration is completed, Partition-Based Sync client apps can communicate
with your App's Flexible Sync backend. You can disable and re-enable Flexible Sync,
but if you re-enable Partitition-Based Sync you will need to do another migration
to enable Flexible Sync later on.

Migrate Permissions and Rules
`````````````````````````````

Your App's permissions and rules will be automatically migrated if the Partition-Based
Sync rules have direct App Services Rules equivalents. This will override
previously-defined App Services Rules.

Some Partition-Based Sync rule strategies can't translate directly to App Services
Rules. You may need to manually migrate rules with these expansions:

- Rules that include :json-expansion:`%%partition`. 
  
  App Services Rules do not have an equivalent expansion.
  :ref:`Function rules <pbs-function-rules>` used with :json-expansion:`%%partition`

>   cannot be migrated.

Author: Oh, I misunderstood. I thought `%function` would work, but not in conjunction with `%%partition`. Thank you for the clarification. I'll update this section to account for that.

commit the migration to make it permanent or revert the migration to return your
App to Partiton-Based Sync.

If you don't commit or revert the migration, it will automatically commit
according to your 
:atlas:`minimum oplog window </atlas/cluster-additional-settings/#set-minimum-oplog-window>`. 
For example, if your minimum oplog window is 48 hours, you will have 48 hours to
revert your migration.

After a migration is committed, the old Partition-Based Sync metadata is deleted
and you cannot revert the migration.

**To manually commit a migration:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select the three dots next to 
   :guilabel:`Manage Sync Migration`.
#. From the list of options, select 
   :guilabel:`Delete Partition-Based Sync metadata`.
   
   .. image:: /images/commit-migration.png
      :alt: Commit a migration
   
   This will commit your migration and you will not be able to recover your
   Partition-Based Sync Metadata.

.. _realm-sync-migrate-revert:

Revert Migration
~~~~~~~~~~~~~~~~

You can revert a migration after it has completed and before it has been committed.

> Completed migrations are automatically committed when the PBS translator falls 

Author: Thank you! I thought I fixed that, but clearly left it in. We do not abbreviate to PBS in the docs.

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

Migrating an App that uses Partition-Based Sync to Flexible Sync is an automatic

> process. Other than upgrading the SDK version, migrating sync modes doesn't require 

Author: Applied locally with other small edits. Will be in next push.

any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend.

Requirements
------------

App Services App:

- Must currently use Partition-Based Sync

Linked Atlas Cluster:

- Must be running MongoDB 5.0+ for Flexible Sync compatibility
- Must have storage auto-scaling enabled
- Must be M10+ to access this setting
- Must have time-based oplog setting enabled

Client apps:

- ``realm-core`` vTBD+, which means these SDK versions: TBD

.. _realm-sync-migrate-what-to-know:

> What to Know Before migrating

Author: Applied locally with other small edits. Will be in next push.

   proportional to the amount of data in synced collections.

#. **Completed migration** 

   :guilabel:`Manage Sync Migration` appears on the Device Sync page if the 
   migration hasn't been committed or reverted.

Migrate Partition-Based Sync to Flexible Sync
---------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

During migration, client apps may experience side effects, most of which are similar
to what happens if you :ref:`terminate <terminating-realm-sync>` and re-enable
Device Sync. These are some additional side effects of migration:

- In the App Services UI, the Rules and Schema pages will be read-only.
- Your App will experience a period of read-only sync and then a 

>   period of downtime.

Reviewer 2: Do we know how long the downtime might be?

Author: Unfortunately, no. And it's hard to give an estimate without potentially putting ourselves in a difficult situation when we're inevitably wrong. I could repeat what I have in the "What to Know Before Migrating" section:
> How long a migration takes is directly proportional to the amount of data that
   needs to be migrated. The more data, the longer it will take to migrate.

Do you think that would be beneficial?

Reviewer 3: I see you added that note above. I don't think it needs to be in two places on the same page.

Linked Atlas Cluster:

- Use MongoDB 5.0 or later for Flexible Sync compatibility
- Have storage auto-scaling enabled
- Be an M10+ Cluster Tier
- Have time-based oplog setting enabled

Client apps:

- Use ``realm-core`` vTBD+, which means these SDK versions: TBD

.. _realm-sync-migrate-what-to-know:

What to Know Before Migrating
-----------------------------

Migrating your App's Sync mode involves changes that affect both client apps and
your App's backend. You should be aware of the side effects and plan ahead.

- How long a migration takes is directly proportional to the amount of data that
  needs to be migrated. The more data, the longer it will take to migrate.
- Migrations can be :ref:`canceled <realm-sync-migrate-cancel>` during the 
  migration or :ref:`reverted <realm-sync-migrate-revert>` within a certain 
  timeframe after the migration is finished.
- Progress notifications will not work after migration because Flexible Sync does
  not support them.
- Client apps that use an SDK with with a realm-core version lower than TBD will
  experience a client reset after migration. You should use a 
  :ref:`client reset <handle-a-client-reset>` handler with recovery enabled.

> - :ref:`Development Mode <development-mode>` is not supported during a migration.

Reviewer: This is a bit misleading - dev mode can still be enabled on the backend, and client apps can still add new columns to existing tables without issues. The scenario that @michael-wb and @danieltabacaru were saying would not work properly is if a client tried to create a new table against an FLX backend while using a PBS config locally.

I don't think it's worth getting into all of the nuances about what exact behavior they will see, so maybe something simple like `Creating new tables via Development Mode may lead to unexpected behavior while using a partition-based sync configuration in your app`

Author: Ha! Yeah, my first local revision for this had a ton of detail. I decided to pare it back and went too far. Thanks.

   :guilabel:`Start Migration`.

.. note:: Temporary Sync Disruption

   Clients will not be able to sync their local writes until the migration has
   completed.

After a migration is completed, Partition-Based Sync client apps can communicate
with your App's Flexible Sync backend. You can disable and re-enable Flexible Sync
and client apps will still communicate with your backend.

If you re-enable Partitition-Based Sync you will lose the "migrated" status.
This means you'll need to to another migration if you want to use Flexible Sync
on the backend with Partition-Based Sync client apps.

Migrate Permissions and Rules
`````````````````````````````

Your App's permissions and rules will be automatically migrated if the Partition-Based
Sync rules have direct App Services Rules equivalents. This will override
previously-defined App Services Rules.

In step 4 above, you will see if your permissions and rules can be automatically
migrated. On the :guilabel:`Review Migration Process` step in the App Services UI,
look for :guilabel:`Rules & Permisssions`.

Some Partition-Based Sync rule strategies can't translate directly to App Services

> Rules. You may need to manually migrate rules that include:

Reviewer: [nit] "rules" -> "permissions" might be more familiar for a PBS user

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

Migrating an App that uses Partition-Based Sync to Flexible Sync is an automatic
process. Other than upgrading the SDK version, migrating Sync modes doesn't require 
any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend. We recommend eventually migrating your client apps to also
use Flexible Sync instead of Partition-Based Sync.

Requirements
------------

App Services App:

- Currently use Partition-Based Sync

Linked Atlas Cluster:

- Use MongoDB 5.0 or later for Flexible Sync compatibility
- Have storage auto-scaling enabled
- Be an M10+ Cluster Tier
- Have time-based oplog setting enabled

Client apps:

> - Use ``realm-core`` vTBD+, which means these SDK versions: TBD

Author: `realm-core` version possibly 13.8.0.

Reviewer 3: Update before deploying!

Author: Thanks for the reminder! We still don't know what Core version (or SDK versions) will be required yet.

Migration Stages
~~~~~~~~~~~~~~~~

There are three stages to a Sync mode migration. You can see your migration's
stage in the App Services UI.

#. **Syncing Partition-Based Sync metadata** 
   
   At this stage, client apps can connect to your App's backend, but local writes 
   will not sync. These writes will be lost if you don't have a 
   :ref:`client reset handler <handle-a-client-reset>` to recover unsynced changes.
	 
   Clients can also receive updates to synced collections, but they won't
   be able to receive upload acknowledgements.

#. **Building Flexible Sync metadata** 
   
   Clients cannot connect at this stage. The duration of this stage is directly
   proportional to the amount of data in synced collections.

#. **Completed migration** 

   :guilabel:`Manage Sync Migration` appears on the Device Sync page if the 
   migration hasn't been committed or reverted.

After your backend is migrated to Flexible Sync, the backend automatically
injects a ``partitionKey`` field into every document the Partition-Based Sync
client creates and sets its value to your sync configuration's
``partitionValue``. The injected ``partitionKey`` field does not sync back to

> client devices. This means that clients won't be able to read this field.

Author: Ooh, that's a good thing to clarify. Thank you!

You cannot revert a migration that has been committed.

Similar to canceling a migration, if you revert a migration, you need to 
start over from the beginning with future migration attempts.

**To revert a migration:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select the three dots next to 
   :guilabel:`Manage Sync Migration`.
#. From the list of options, select 
   :guilabel:`Revert back to Partition-Based Sync`.
   
   .. image:: /images/revert-migration.png
      :alt: Revert a migration
   
   This will revert your migration. You will need to migrate again if you 
   want to enable Flexible Sync in the future.

.. _realm-sync-migrate-client:

Migrate Client App to Flexible Sync
-----------------------------------

After your app's backend has been migrated to Flexible Sync, we recommend
that you also migrate your client app to Flexible Sync. When you migrate your
client app to Flexible Sync, the client will stop automatically creating
sync subscriptions for each Realm object model in your app's data model.

We also recommend removing all sync subscriptions and then creating new subscriptions

> for your data. This is the clearest way to sync data after a migration.

Reviewer: hmm, not sure i get what this second sentence means. IIUC, maybe "This is the clearest way to control what data is synced after a migration" ?

Author: Yes, thank you for pointing out the ambiguity here.

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

> Migrating an App that uses Partition-Based Sync to Flexible Sync is an automatic

Reviewer 3: I think "app" should be lowercase here.

Reviewer 3: And are we talking about a client app or AAS app?

Author: Ah, good point. I'll make this clearer.

> process. Other than upgrading the SDK version, migrating Sync modes doesn't require 

Reviewer 3: Is upgrading the SDK required? Seems like only if it is old-ish. As I read the sentence, it implies one must upgrade. If not, perhaps: Other than possibly needing to upgrade the SDK version...

Author: It is required. There's some automatic logic added to the SDK to enable the PBS client to communicate with the Flexible Sync backend. SDK versions older than the (TBD) version that includes the logic simply won't be able to communicate with the backend.

any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend. We recommend eventually :ref:`migrating your client app 
<realm-sync-migrate-client>` to also use Flexible Sync instead of
Partition-Based Sync.

Requirements
------------

App Services App:

- Currently use Partition-Based Sync

Linked Atlas Cluster:

- Use MongoDB 5.0 or later for Flexible Sync compatibility
- Have storage auto-scaling enabled
- Be an M10+ Cluster Tier
- Have time-based oplog setting enabled

Client apps:

- Use ``realm-core`` vTBD+, which means these SDK versions: TBD

.. _realm-sync-migrate-what-to-know:

What to Know Before Migrating
-----------------------------

> Migrating your App's Sync mode involves changes that affect both client apps and

Reviewer 3: App => app throughout

your App's backend. You should be aware of the side effects and plan ahead.

> - Client apps that use an SDK with with a realm-core version lower than TBD will

Reviewer 3: Watch out for the "TBD"s!!!

Author: Those TBDs'll get ya every time. 😆 

I'll update when we've got better version info.

  not be able to sync after migration.
- How long a migration takes is directly proportional to the amount of data that
  needs to be migrated. The more data, the longer it will take to migrate.
- Migrations can be :ref:`canceled <realm-sync-migrate-cancel>` during the 
  migration or :ref:`reverted <realm-sync-migrate-revert>` within a certain 
  timeframe after the migration is finished.
- Progress notifications will not work after migration because Flexible Sync does
  not support them.  
- Client apps that were connected to the Partition-Based Sync backend will 
  experience a client reset after the backend is migrated to Flexible Sync. You 
  should use a :ref:`client reset <handle-a-client-reset>` handler with recovery enabled.
- Creating new tables in a migrated Flexible Sync backend using 
  :ref:`Development Mode <development-mode>` from a Partition-Based Sync client
  may lead to unexpected behavior.

.. _realm-sync-migrate-stages:

Migration Stages
~~~~~~~~~~~~~~~~

> There are three stages to a Sync mode migration. You can see your migration's

Reviewer 3: Consider switching the order of these first two sentences. That way, the second one is a lead-in to the list that follows. Might even consider ending it with a colon (:)

Author: Excellent suggestion. Thank you!

   be able to receive upload acknowledgements.

#. **Building Flexible Sync metadata** 
   
   Clients cannot connect at this stage. The duration of this stage is directly
   proportional to the amount of data in synced collections.

#. **Completed migration** 

   :guilabel:`Manage Sync Migration` appears on the Device Sync page if the 
   migration hasn't been committed or reverted.

After the migration has finished, check out the :ref:`After Migration
<realm-sync-migrate-after>` section for more information about how
Partition-Based Sync clients interact with the Flexible Sync backend.

Migrate Partition-Based Sync to Flexible Sync
---------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

> During migration, client apps may experience side effects, most of which are similar

Reviewer 3: "Side effects may include painful rash..." ;-) 
Seriously, I always think of "side effects" in that way, and I wonder if others do, too? Also, do client apps "experience" anything? (not to be too philosophical here...)

If so, consider "client apps may be impacted similar to what happens when you terminate (etc)"

Author: Ah, there I go anthropomorphizing technology again. I do hope every realm has a beautiful experience though. 😄 

I'll re-word this.

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

Migrating an App Services App that uses Partition-Based Sync to Flexible Sync is
an automatic process. Other than upgrading the SDK version, migrating Sync modes
doesn't require any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend. We recommend eventually :ref:`migrating your client app 
<realm-sync-migrate-client>` to also use Flexible Sync instead of
Partition-Based Sync.

Requirements
------------

App Services App:

- Currently use Partition-Based Sync

Linked Atlas Cluster:

- Use MongoDB 5.0 or later for Flexible Sync compatibility
- Have storage auto-scaling enabled
- Be an M10+ Cluster Tier
- Have time-based oplog setting enabled

Client apps:

> - Use ``realm-core`` vTBD+, which means these SDK versions: TBD

Reviewer: ```suggestion
- Use ``realm-core`` v13.10.0+, which means these SDK versions: TBD
```

.. _realm-sync-migrate-what-to-know:

What to Know Before Migrating
-----------------------------

Migrating your App's Sync mode involves changes that affect both client apps and
your App's backend. You should be aware of the effects and plan ahead.

> - Client apps that use an SDK with with a ``realm-core`` version lower than TBD will

Reviewer: ```suggestion
- Client apps that use an SDK with with a ``realm-core`` version lower than v13.10.0 will
```

   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync
and Flexible Sync. Partition-Based Sync is an older mode and you should consider
migrating to Flexible Sync.

Migrating an App Services App that uses Partition-Based Sync to Flexible Sync is
an automatic process. Other than upgrading the SDK version, migrating Sync modes
doesn't require any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend. We recommend eventually :ref:`migrating your client app 
<realm-sync-migrate-client>` to also use Flexible Sync instead of
Partition-Based Sync.

Requirements
------------

Currently, the Kotlin and Swift SDKs can migrate from Partition-Based Sync to
Flexible Sync. The C++, Flutter, Java, .NET, Node.js, and React Native SDKs
cannot perform the migration.

App Services App:

- Currently use Partition-Based Sync

Linked Atlas Cluster:

- Use MongoDB 5.0 or later for Flexible Sync compatibility
- Have storage auto-scaling enabled
- Be an M10+ Cluster Tier
- Have time-based oplog setting enabled

Client apps:

> - ``realm-core`` v13.10.0 or later, which is included in these SDK versions:

Author: Good point. No need to make devs think about realm-core.

  - Realm Kotlin SDK v1.8.0
  - Realm Swift SDK v10.39.0

.. _realm-sync-migrate-what-to-know:

What to Know Before Migrating
-----------------------------

Migrating your App's Sync mode involves changes that affect both client apps and
your App's backend. You should be aware of the effects and plan ahead.

- Client apps that use an SDK with with a ``realm-core`` version earlier than
  v13.10.0 will not be able to sync after migration.
- How long a migration takes is directly proportional to the amount of data that
  needs to be migrated. The more data, the longer it will take to migrate.
- Migrations can be :ref:`canceled <realm-sync-migrate-cancel>` during the 
  migration or :ref:`reverted <realm-sync-migrate-revert>` within a certain 
  timeframe after the migration is finished.

> - Progress notifications will not work after migration because Flexible Sync does

Author: @kmorkos, would having client code for progress notifications cause fatal client errors? I know one of the goals of the project was to prevent devs from needing to make client code changes. Is this something that has to be done?

> - Progress notifications will not work after migration because Flexible Sync does

Reviewer: It won't trigger any errors, the handlers will just always report 100% progress. So they can keep it in, it will just be unreliable

> - Progress notifications will not work after migration because Flexible Sync does

Author: I think I'll leave this where it is, then. It's not great to have unnecessary client code, but removing it isn't exactly required.

Partition-Based Sync clients interact with the Flexible Sync backend.

Migrate Partition-Based Sync to Flexible Sync
---------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

Migration generally has similar effects as :ref:`terminating <terminating-realm-sync>` 
and re-enabling Device Sync. There are some additional effects of migration:

- In the App Services UI, the Rules and Schema pages will be read-only.
- Your App will experience a period of read-only Sync and then a 
  period of downtime.
- The new Flexible Sync configuration will set 
  :ref:`client max offline time <client-maximum-offline-time>` to 30 days.
- Storage in your Atlas cluster used by Device Sync metadata is temporarily doubled.

>   Partition-Based Sync metadata and Flexible Sync metadata must temporarily co-exist. 

Reviewer 4: Is there any information about _how long_ this metadata must temporarily co-exist? i.e. is it for the client max offline time or some other definitive period, or is it variable?

Author: Good point. It only lasts as long as the migration itself. As soon as the migration is rolled back or completed, the doubled metadata is removed.

Author: Actually, @kmorkos, what happens if there isn't enough space in the cluster to double the metadata? Would this be a rare edge case?

Author: Ahhh, that makes perfect sense. Thank you!

  Depending on the size of your synced data, this could affect billing every time you
  migrate.

**To migrate from Partition-Based Sync to Flexible Sync:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select :guilabel:`[Migrate to Flexible Sync]`.
   
   .. image:: /images/begin-migration.png
      :alt: Begin a migration
   
#. Read the information about migrating, then select 
   :guilabel:`Next: Migration Requirements`.
#. Ensure you meet the migration requirements, the select 
   :guilabel:`Start Migration`.

.. note:: Temporary Sync Disruption

   Clients will not be able to sync their local writes until the migration has
   completed.

> After a migration is completed, Partition-Based Sync client apps can communicate

Reviewer 4: You mentioned above that clients must perform a client reset after migrating. Maybe it's worth adding that here?

and client apps will still communicate with your backend.

If you re-enable Partitition-Based Sync you will lose the "migrated" status.
This means you'll need to to another migration if you want to use Flexible Sync
on the backend with Partition-Based Sync client apps.

Migrate Permissions and Rules
`````````````````````````````

Your App's permissions and rules will be automatically migrated if the Partition-Based
Sync rules have direct App Services Rules equivalents. This will override
previously-defined App Services Rules.

In step 4 above, you will see if your permissions and rules can be automatically
migrated. On the :guilabel:`Review Migration Process` step in the App Services UI,
look for :guilabel:`Rules & Permisssions`.

Some Partition-Based Sync rule strategies can't translate directly to App Services
Rules. You may need to manually migrate permissions that include:

- :json-operator:`%function` operator.

  :ref:`Function rules <pbs-function-rules>` are not compatible with Flexible Sync 
  and cannot be migrated.

- :json-expansion:`%%partition` expansion.

  App Services Rules do not have an equivalent expansion for  

>   :json-expansion:`%%partition`, so it cannot be migrated.

Reviewer: we actually do migrate this automatically if it's not used in conjunction with the other restrictions listed here.
e.g. `{ %%partition: %%user.id }` becomes `{ <partitionKey>: %%user.id }`

if they need to manually update their permissions, then this is implied in the "See the list of Flexible Sync-compatible expansions for all supported expansions." note below so we can probably remove the bullet here, or change up the wording/move it elsewhere in this section

- ``%or``, ``%not``, ``%nor``, ``%and`` expansions.

  These permissions may work, but there is enough nuance that you should test them
  to ensure expected behavior. Testing new permissions won't work on the App

>   you are migrating. You'll need a new app to test manually-migrated permissions.

Reviewer 4: There's a little ambiguity here. Are we advising people to test manually-migrated permissions in a new FS app, and _then_ apply them manually here so they can complete the migration?

Author: Yeah, this section was tricky to write. It's a "probably will work, but we can't guarantee it" situation. And, yes, your read is correct. I'll edit to make that clearer.

Reviewer: The divergence is mainly with the permission behavior in non-sync services. An `%or` expression in an app services rule that works in flexible sync will not work outside of sync and vice versa 🥴. This is a known issue that's currently being addressed, so once that's fixed we _may_ be able to migrate those scenarios automatically (can't promise that though!)

Similar to canceling a migration, if you revert a migration, you need to 
start over from the beginning with future migration attempts.

**To revert a migration:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select the three dots next to 
   :guilabel:`Manage Sync Migration`.
#. From the list of options, select 
   :guilabel:`Revert back to Partition-Based Sync`.
   
   .. image:: /images/revert-migration.png
      :alt: Revert a migration
   
   This will revert your migration. You will need to migrate again if you 
   want to enable Flexible Sync in the future.

.. _realm-sync-migrate-after:

After Backend Migration
-----------------------

Your migrated backend uses a partition key field in your Realm object models
to map the Partition-Based Sync client objects to Flexible Sync equivalents on
the backend. If your object models don't have a partition key field, the
backend automatically injects one into every object the Partition-Based Sync
client creates.

These are some other characteristics of the injected field you should be aware
of:

- The injected field follows this structure: ``<yourPartitionKey> == 

>   <yourPartitionValue>``. Both the field key is derived from your old

Reviewer 4: ```suggestion
  <yourPartitionValue>``. The field key is derived from your old
```

Author: Thanks for catching this. "Both the field key" just rolls of the tongue though... 😆

> client creates.

Reviewer 5: You might want to point out that this value will no longer be injected once the client app has been migrated to flexible sync, which could lead to the PBS client apps not seeing the new data. The workaround is to include the partition key field in the object model and manually fill in this value if the data should be seen by PBS client apps.

Author: @michael-wb, I'm not sure I follow.
> once the client app has been updated to use flexible sync
> which could lead to the PBS client apps not seeing the new data

Both can't be true, can they? The PBS clients should update to Flexible Sync.

Reviewer 5: Hopefully this is clearer:

When a PBS client app has migrated to use flexible sync, the partition field will be populated with the partition value automatically by the server. The PBS client app in the migrated state relies on this partition value to subscribe to the data to download from the server.
One the client app has been updated to natively use flexible sync (i.e., no longer a PBS app), the server will no longer automatically populate the partition field. Any PBS versions of the client app still out in the wild will not see/download any objects created by the FLX version of the client app unless the partition field is manually populated when the object is created.

Reviewer 5: It is not guaranteed that users will update their apps (at the same time or ever) to the new FLX version.

Author: OK, I was wondering if that was the case. Thank you for clarifying!

  Partition-Based Sync  **backend**  configuration. The value is derived from
  your current Partition-Based Sync  **client**  configuration.
- The injected field does not sync back to the client that created the object.
  This means the originating client won't be able to read the injected
  ``partitionKey`` field.

The Partition-Based Sync client automatically creates sync subscriptions for 
each table where ``partitionKey == partitionValue``. This will continue
until you :ref:`migrate your client app <realm-sync-migrate-client>` to
Flexible Sync.

.. _realm-sync-migrate-client:

Migrate Client App to Flexible Sync
-----------------------------------

After your app's backend has been migrated to Flexible Sync, we recommend

> that you also migrate your client app to Flexible Sync. When you migrate your

Reviewer 4: In the SDK side, we should spell out exactly what it means to "migrate your client app to Flexible Sync." I assume that's something like:

- Update a Partition-Based Configuration to a Flexible Sync configuration
- Keep, remove, or update the automatically-generated subscriptions as needed (the Core PR says subscriptions will be given a name in the format "flx_migrated_")
- Add any fields models need to manage subscriptions. i.e. "If your object models don't have a partition key field, the backend automatically injects one into every object the Partition-Based Sync client creates... The injected field does not sync back to the client that created the object." Would we advise devs to specifically add such a key to their models so they're visible and can be updated as needed? And so they remember to give new models the relevant key?

Author: Yeah, I think this info will be helpful for the SDK docs. I've captured this info in the Jira ticket for the SDK docs work.

Migrate Client App to Flexible Sync
-----------------------------------

After your app's backend has been migrated to Flexible Sync, we recommend
that you also migrate your client app to Flexible Sync. When you migrate your
client app to Flexible Sync, the client will stop automatically creating
sync subscriptions for each Realm object model in your app's data model.

We also recommend removing all sync subscriptions and then creating new subscriptions
for your data. This is the clearest way to control what data is synced after a
migration. Refer to the SDK documentation for details about adding and removing

> subscriptions.

Reviewer 5: I don't know if you want to mention the naming format of the generated subscriptions while the client is in the migrated state. It follows this format: `flx_migrated_` followed by the name of the table/collection. For example, a subscription for the "Person" collection will be named: `flx_migrated_Person`

   :ref:`Manage Flexible Sync Subscriptions - Flutter SDK <flutter-flexible-sync-manage-subscriptions>`

- Java SDK
   :ref:`Flexible Sync - Java SDK <java-flexible-sync>`

- Kotlin SDK
   :ref:`Subscribe to Queryable Fields - Kotlin SDK <kotlin-subscriptions>`

- .NET SDK
   :ref:`Manage Flexible Sync Subscriptions - .NET SDK <dotnet-flexible-sync>`

- Node.js SDK
   :ref:`Flexible Sync - Node.js SDK <node-flexible-sync>`

- React Native SDK
   :ref:`Flexible Sync - React Native SDK <react-native-flexible-sync>`

- Swift SDK
   :ref:`Manage Flexible Sync Subscriptions - Swift SDK <swift-manage-flexible-sync-subscriptions>`

If you don't remove all subscriptions, the automatically-created subscriptions

> will remain, but you won't see them in your client code. This makes it harder

Reviewer 4: "...but you won't see them in your client code" was a little confusing to me. I assumed they would be available in the realm.subscriptions set with the autogenerated names along with any other subscriptions. But after re-reading this I think you mean something like "you don't see a code path in your code base that creates these subscriptions so you might forget these exist." (Which is probably also true of the partition key we automagically create.)

I don't have a specific suggestion, but I wonder if it's worth trying to make clearer that these subscriptions are accessible on the client but because they weren't created in the client codebase, it might make maintaining and extending client code confusing.

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: Partition-Based Sync

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Author: We use "is an older mode" in other places throughout the docs. I think we were hedging our bets a bit because we weren't sure what product's EOL plan for Partition-Based Sync is. Has it been communicated anywhere other than docs that PBS may be deprecated?

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Reviewer 5: What about "Partition-Based Sync may be deprecated in the future and you should..."

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Reviewer 5: Should you also include ref links for more information about the two modes?

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Author: I don't think docs are a good place to mention that PBS may be deprecated in the future. Unless product has an EOL plan I haven't heard of (entirely possible), it's pure speculation at this point.

Good point about adding ref links. I'll add those.

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Author: I can follow up with Ian when he's back from vacation.

> and Flexible Sync. Partition-Based Sync is an older mode and you should consider

Reviewer: We do have a ticket to disallow enabling PBS on new sync apps which effectively makes it "deprecated", but I agree this is maybe not the right place for users to be hearing this for the first time

  - Realm Swift SDK v10.39.0
- Have a :ref:`Client Reset Handler <client-resets>` configured.
- Remove any code in a client app that relies on progress notifications because Flexible Sync does not support them

.. _realm-sync-migrate-what-to-know:

What to Know Before Migrating
-----------------------------

Migrating your App's Sync mode involves changes that affect both client apps and
your App's backend. You should be aware of the effects and plan ahead.

- Client apps that don't meet the minimum SDK version requirement will not be
  able to sync after migration.
- How long a migration takes is directly proportional to the amount of data that
  needs to be migrated. The more data, the longer it will take to migrate.
- Migrations can be :ref:`canceled <realm-sync-migrate-cancel>` during the 
  migration or :ref:`reverted <realm-sync-migrate-revert>` within a certain 
  timeframe after the migration is finished.
- Progress notifications will not work after migration because Flexible Sync does
  not support them.  
- Client apps that were connected to the Partition-Based Sync backend will 
  experience a client reset after the backend is migrated to Flexible Sync. You 

>   should use a :ref:`client reset <handle-a-client-reset>` handler with recovery enabled.

Reviewer 5: ... with recovery enabled to avoid losing any pending changes from before the client reset.

- Creating new tables in a migrated Flexible Sync backend using 
  :ref:`Development Mode <development-mode>` from a Partition-Based Sync client

>   may lead to unexpected behavior.

Reviewer 5: Or making changes to the current schema?

Possible update:
Making changes to the current schema or creating new tables using :ref:``Development Mode <development-mode>`` in a migrated Flexible Sync backend may lead to unexpected behavior for the Partition-Based Sync clients.

Reviewer: IIRC this was specifically about the PBS client not creating the subscriptions on the new tables, so the only relevant schema change here is adding a new table. I think the current wording is correct.

.. _realm-sync-migrate-stages:

Migration Stages
~~~~~~~~~~~~~~~~

You can see your migration's stage in the App Services UI. There are three stages
to a Sync mode migration:

#. **Syncing Partition-Based Sync metadata** 
   
   At this stage, client apps can connect to your App's backend, but local writes 
   will not sync. These writes will be lost if you don't have a 
   :ref:`client reset handler <handle-a-client-reset>` to recover unsynced changes.
	 

>    Clients can also receive updates to synced collections, but they won't

Reviewer 6: I'd say "clients will receive updates to synced collections, but won't be able to send uploads or receive upload acknowledgements".

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

Migration generally has similar effects as :ref:`terminating <terminating-realm-sync>` 
and re-enabling Device Sync. There are some additional effects of migration:

- In the App Services UI, the Rules and Schema pages will be read-only.
- Your App will experience a period of read-only Sync and then a 
  period of downtime.
- The new Flexible Sync configuration will set 
  :ref:`client max offline time <client-maximum-offline-time>` to 30 days.
- Storage in your Atlas cluster used by Device Sync metadata is temporarily doubled.
  Partition-Based Sync metadata and Flexible Sync metadata must temporarily co-exist. 
  Depending on the size of your synced data, this could affect billing every time you

>   migrate. The doplicated metadata is removed after the migration is rolled back

Reviewer 6: duplicated*

  or committed.

**To migrate from Partition-Based Sync to Flexible Sync:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select :guilabel:`[Migrate to Flexible Sync]`.
   
   .. image:: /images/begin-migration.png
      :alt: Begin a migration
   
#. Read the information about migrating, then select 
   :guilabel:`Next: Migration Requirements`.
#. Ensure you meet the migration requirements, the select 
   :guilabel:`Start Migration`.

.. note:: Temporary Sync Disruption

   Clients will not be able to sync their local writes until the migration has
   completed.

After a migration is completed, Partition-Based Sync client apps will go through
a client reset. Then, they can communicate with your App's Flexible Sync backend.
You can disable and re-enable Flexible Sync and client apps will still
communicate with your backend.

If you re-enable Partitition-Based Sync you will lose the "migrated" status.

> This means you'll need to to another migration if you want to use Flexible Sync

Reviewer 6: "to perform", or "to do"

for all supported expansions.

You should also check out the 
:ref:`Device Sync Permissions Guide <flexible-sync-permissions-guide>` for more
information about how to work with permissions.

.. _realm-sync-migrate-cancel:

Cancel Migration
~~~~~~~~~~~~~~~~

You can cancel a migration at any time while it is in progress. This will return
your App the Partition-Based Sync, with all settings as they were before the
migration started.

If you cancel a migration, you need to start over from the beginning with future
migration attempts.

**To cancel a migration:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, in the notification banner, select
   :guilabel:`Stop Migration`.
   
   .. image:: /images/cancel-migration.png
      :alt: Cancel a migration
   
   This will cancel your migration. You will need to migrate again if you want to
   enable Flexible Sync in the future.

.. _realm-sync-migrate-commit:

Commit Migration
~~~~~~~~~~~~~~~~

After your migration has completed, it is in an evaluation state. You can manually
commit the migration to make it permanent or revert the migration to return your

> App to Partiton-Based Sync.

Reviewer 6: partition**

Author: Nice catch! Thanks for pointing this out.

.. _realm-sync-migrate-after:

After Backend Migration
-----------------------

Your migrated backend uses a partition key field in your Realm object models
to map the Partition-Based Sync client objects to Flexible Sync equivalents on
the backend. If your object models don't have a partition key field, the
backend automatically injects one into every object the Partition-Based Sync
client creates.

These are some other characteristics of the injected field you should be aware
of:

- The injected field follows this structure: ``<yourPartitionKey> == 
  <yourPartitionValue>``. The field key is derived from your old
  Partition-Based Sync  **backend**  configuration. The value is derived from
  your current Partition-Based Sync  **client**  configuration.
- The injected field does not sync back to the client that created the object.
  This means the originating client won't be able to read the injected
  ``partitionKey`` field.

The Partition-Based Sync client automatically creates sync subscriptions for 
each table where ``partitionKey == partitionValue``. This will continue
until you :ref:`migrate your client app <realm-sync-migrate-client>` to
Flexible Sync.

> After migrating your client app to Flexible Sync, the backend will stop injecting

Reviewer 6: I'd say "client app code" again here. It's also not super clear what this means I think - maybe we can say "objects created in new versions of your app will only be available to older versions if you include the `partitionKey` field" (needs some wordsmithing but something like that)

the partition key field. This means objects created after migrating your client
app to Flexible Sync may not be available in clients that haven't yet been
updated to Flexible Sync.

.. _realm-sync-migrate-client:

> Migrate Client App to Flexible Sync

Reviewer 6: again maybe "Client App Code"

-----------------------------------

After your app's backend has been migrated to Flexible Sync, we recommend
that you migrate your client app to Flexible Sync. When you migrate your
client app to Flexible Sync, the client will stop automatically creating
sync subscriptions for each Realm object model in your app's data model.

We also recommend removing all sync subscriptions and then creating new subscriptions
for your data. This is the clearest way to control what data is synced after a
migration. If you want to remove individual subscriptions, the automatically

> generated subscriptions use this naming format: ``flx_migrated_<objecModelName>``.

Reviewer 5: objectModelName?

Author: The `<placeHolder>` bit is how we indicate placeholder variable text in the docs. The idea being the tables are created based on client object model names, right? Maybe a specific example, is less confusing.

I'll take another pass on this.

Reviewer: i assume @michael-wb is calling out the typo here? it's a sneaky one
```suggestion
generated subscriptions use this naming format: ``flx_migrated_<objectModelName>``.
```

Reviewer 5: Sorry - your text says `<objecModelName>`, I wasn't sure if you meant to not include the `t`...

Author: oooooh. lmao. I've rewritten to make it clearer anyway. 😆

For example, a subscription for the ``Person`` collection would be named
``flx_migrated_Person``.

Refer to the SDK documentation for details about adding and removing
subscriptions:

- Java SDK
   :ref:`Flexible Sync - Java SDK <java-flexible-sync>`

- Kotlin SDK
   :ref:`Subscribe to Queryable Fields - Kotlin SDK <kotlin-subscriptions>`

- .NET SDK
   :ref:`Manage Flexible Sync Subscriptions - .NET SDK <dotnet-flexible-sync>`

- Node.js SDK
   :ref:`Flexible Sync - Node.js SDK <node-flexible-sync>`

- React Native SDK
   :ref:`Flexible Sync - React Native SDK <react-native-flexible-sync>`

- Swift SDK
   :ref:`Manage Flexible Sync Subscriptions - Swift SDK <swift-manage-flexible-sync-subscriptions>`

You can keep the automatically-generated subscriptions if you don't want to
remove and recreate them. But you should consider that because these 
subscriptions aren't created in your client code, they may make it harder to

> maintain and extend client code.

Reviewer 5: And they will not exist on any clients that never connected to the App after the migration to Flexible Sync was performed.

.. _realm-sync-migrate-modes:

=========================
Migrate Device Sync Modes
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Atlas App Services Apps can have one of two Device Sync modes: 
:ref:`Partition-Based Sync <partition-based-sync>` and :ref:`Flexible Sync <sync>`.
Partition-Based Sync is an older mode and you should consider migrating to
Flexible Sync.

Migrating an App Services App that uses Partition-Based Sync to Flexible Sync is
an automatic process. Other than upgrading the SDK version, migrating Sync modes
doesn't require any changes to your client app code.

Migration results in Partition-Based Sync clients communicating with a
Flexible Sync backend. We recommend eventually :ref:`migrating your client app code
<realm-sync-migrate-client>` to also use Flexible Sync instead of
Partition-Based Sync.

Requirements
------------

> Currently, the Kotlin, SDK supports migrating your App's backend from

Reviewer: ```suggestion
Currently, the Kotlin SDK supports migrating your App's backend from
```

   proportional to the amount of data in synced collections.

#. **Completed migration** 

   :guilabel:`Manage Sync Migration` appears on the Device Sync page if the 
   migration hasn't been committed or reverted.

After the migration has finished, check out the :ref:`After Migration
<realm-sync-migrate-after>` section for more information about how
Partition-Based Sync clients interact with the Flexible Sync backend.

Migrate Partition-Based Sync to Flexible Sync
---------------------------------------------

After beginning migration from Partition-Based Sync to Flexible Sync, the process
is mostly automatic. However, you can control important parts of the migration.

During a migration, you can cancel it. After a migration, you can revert it 
or manually commit the migration.

Begin Migration
~~~~~~~~~~~~~~~

Before you begin a migration, make sure you're aware of the
:ref:`migration stages <realm-sync-migrate-stages>` and
:ref:`effects <realm-sync-migrate-what-to-know>`.

Migration generally has similar effects as :ref:`terminating <terminating-realm-sync>` 
and re-enabling Device Sync. There are some additional effects of migration:

> - In the App Services UI, the Rules and Schema pages will be read-only.

Reviewer: not just rules + schema, _all_ config pages will be read-only (functions, auth config, triggers, etc.)

- Your App will experience a period of read-only Sync and then a 
  period of downtime.
- The new Flexible Sync configuration will set 
  :ref:`client max offline time <client-maximum-offline-time>` to 30 days.
- Storage in your Atlas cluster used by Device Sync metadata is temporarily doubled.
  Partition-Based Sync metadata and Flexible Sync metadata must temporarily co-exist. 
  Depending on the size of your synced data, this could affect billing every time you
  migrate. The duplicated metadata is removed after the migration is rolled back
  or committed.

**To migrate from Partition-Based Sync to Flexible Sync:** 

#. In the App Services UI, go to your App and select the Device Sync page.
#. At the top of the page, select :guilabel:`[Migrate to Flexible Sync]`.
   
   .. image:: /images/begin-migration.png
      :alt: Begin a migration
   
#. Read the information about migrating, then select 
   :guilabel:`Next: Migration Requirements`.

> #. Ensure you meet the migration requirements, the select 

Reviewer: ```suggestion
#. Ensure you meet the migration requirements, then select 
```

generated subscriptions use a specific naming format: ``flx_migrated_`` + your 
object model's name. For example, a ``Person`` object model's subscription
name would be ``flx_migrated_Person``.

Refer to the SDK documentation for details about adding and removing
subscriptions:

- Java SDK
   :ref:`Flexible Sync - Java SDK <java-flexible-sync>`

- Kotlin SDK
   :ref:`Subscribe to Queryable Fields - Kotlin SDK <kotlin-subscriptions>`

- .NET SDK
   :ref:`Manage Flexible Sync Subscriptions - .NET SDK <dotnet-flexible-sync>`

- Node.js SDK
   :ref:`Flexible Sync - Node.js SDK <node-flexible-sync>`

- React Native SDK
   :ref:`Flexible Sync - React Native SDK <react-native-flexible-sync>`

- Swift SDK
   :ref:`Manage Flexible Sync Subscriptions - Swift SDK <swift-manage-flexible-sync-subscriptions>`

You can keep the automatically-generated subscriptions if you don't want to
remove and recreate them. But you should consider that because these 
subscriptions aren't created in your client code, they may make it harder to
maintain and extend client code. 

Automatically-generated subscriptions will not exist on clients that never

> connect to your App after it has been migrated to Flexible Sync.

Reviewer: Can't remember the context around why this note was added, but it's unclear why it's important if the client is not re-connecting? Maybe it's referring to a client app that connects as FLX after the migration without ever having connected as PBS? Probably worth clarifying if that's the case

Author: Agreed, it's not clear. I think we could safely remove, as this should be self-evident based on all the behavior described elsewhere on the page.

Requirements
------------

Currently, the Kotlin and Swift SDKs can migrate from Partition-Based Sync to

> Flexible Sync. The C++, Flutter, Java, .NET, Node.js, and React Native SDKs

Reviewer: can remove flutter from this list since migration is irrelevant there. i'm also not sure if the c++ sdk supports PBS? can't find anything in the docs indicating that it does

Reviewer 5: Correct - Flutter and C++ are the newest SDKs and only support flexible sync.

   This will revert your migration. You will need to migrate again if you 
   want to enable Flexible Sync in the future.

.. _realm-sync-migrate-after:

After Backend Migration
-----------------------

Your migrated backend uses a partition key field in your Realm object models
to map the Partition-Based Sync client objects to Flexible Sync equivalents on
the backend. If your object models don't have a partition key field, the

> backend automatically injects one into every table the Partition-Based Sync

Reviewer: ```suggestion
backend automatically injects one into every object the Partition-Based Sync
```

client creates.

The injected field follows this structure: ``<yourPartitionKey> == 
<yourPartitionValue>``. Both the field key and value are derived from your 

> Partition-Based Sync client configuration.

Reviewer: [nit] not sure if this is a distinction worth making, but technically `yourPartitionKey` is a backend parameter, while `yourPartitionValue` is a client config parameter

Author: Ah, thanks for clarifying.
> worth clarifying that the field name will be the partition key from their old backend PBS configuration, not the literal field name "partitionKey"

I guess I missed the "backend" part of what you said earlier. Sorry! I'll update.

There are some other characteristics of the injected field you should be aware of:

> - The injected field does not sync back to the client that created the table.

Reviewer: ```suggestion
- The injected field does not sync back to the client that created the object.
```

  This means the originating client won't be able to read the injected
  ``partitionKey`` field.

> - The Partition-Based Sync client automatically creates sync subscriptions for 

Reviewer: this is not really a characteristic of the injected field, so it should probably be pulled out of this list

Author: Ah, yes. Thank you for noticing this discrepancy. I've re-organized a bit.
